File: android/src/main/kotlin/com/jameskarl/amap/map/bean/CircleOptionsData.kt
package com.jameskarl.amap.map.bean

import com.amap.api.maps.model.CircleHoleOptions
import com.amap.api.maps.model.CircleOptions
import com.jameskarl.amap.map.IJsonEntity

class CircleOptionsData(
        var holeOptions: List<CircleHoleOptionsData>? = null,
        var center: LatLngData? = null,
        var fillColor: Long? = null,
        var radius: Double? = null,
        var strokeColor: Long? = null,
        var strokeDottedLineType: Long? = null,
        var strokeWidth: Float? = null,
        var zIndex: Float? = null,
        var visible: Boolean? = null
) : IJsonEntity {

    fun toCircleOptions(): CircleOptions {
        return CircleOptions().also { option ->
            holeOptions?.let { option.addHoles(it.map { hole -> hole.toBaseHoleOptions() }) }
            center?.let { option.center(it.toLatLng()) }
            fillColor?.let { option.fillColor(it.toInt()) }
            radius?.let { option.radius(it) }
            strokeColor?.let { option.strokeColor(it.toInt()) }
            strokeDottedLineType?.let { option.setStrokeDottedLineType(it.toInt()) }
            strokeWidth?.let { option.strokeWidth(it) }
            zIndex?.let { option.zIndex(it) }
            visible?.let { option.visible(it) }
        }
    }
}

class CircleHoleOptionsData(
        var center: LatLngData? = null,
        var radius: Double? = null
) : IJsonEntity {
    fun toBaseHoleOptions(): CircleHoleOptions {
        return CircleHoleOptions().also { option ->
            center?.let { option.center(it.toLatLng()) }
            radius?.let { option.radius(it) }
        }
    }
}

File: ios/Classes/bean/CircleOptionsData.swift
//
//  CircleOptionsData.swift
//  amap
//
//  Created by James Karl on 2020/4/26.
//

import Foundation

struct CircleOptionsData {
    var holeOptions: [CircleHoleOptionsData]? = nil
    var center: LatLngData? = nil
    var fillColor: Int? = nil
    var radius: Double? = nil
    var strokeColor: Int? = nil
    var strokeDottedLineType: Int? = nil
    var strokeWidth: Float? = nil
    var zIndex: Float? = nil
    var visible: Bool? = nil
    
    init(json: [String:Any?]) {
        self.fillColor = json["fillColor"] as? Int
        self.radius = json["radius"] as? Double
        self.strokeColor = json["strokeColor"] as? Int
        self.strokeDottedLineType = json["strokeDottedLineType"] as? Int
        self.strokeWidth = json["strokeWidth"] as? Float
        self.zIndex = json["zIndex"] as? Float
        self.visible = json["visible"] as? Bool
        
        if let centerValue = json["center"] as? [String: Double] {
            self.center = LatLngData(json: centerValue)
        }
        
        if let holeOptionsValue = json["holeOptions"] as? [Any] {
            if !holeOptionsValue.isEmpty {
                self.holeOptions = []
                holeOptionsValue.forEach { (option) in
                    if let o = option as? [String: Any?] {
                        self.holeOptions?.append(CircleHoleOptionsData(json: o))
                    }
                }
            }
        }
    }
}

struct CircleHoleOptionsData {
    var center: LatLngData? = nil
    var radius: Double? = nil
    
    init(json: [String: Any?]) {
        self.radius = json["radius"] as? Double
        if let centerValue = json["center"] as? [String: Any?] {
            self.center = LatLngData(json: centerValue)
        }
    }
}
